packages/data-stores/src/user/types.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * External dependencies
+ */
+
+import { Action } from 'redux';
+
+export const enum ActionType {
+	CREATE_ACCOUNT = 'CREATE_ACCOUNT',
+	RECEIVE_CURRENT_USER = 'RECEIVE_CURRENT_USER',
+	RECEIVE_CURRENT_USER_FAILED = 'RECEIVE_CURRENT_USER_FAILED',
+	RECEIVE_NEW_USER = 'RECEIVE_NEW_USER',
+	RECEIVE_NEW_USER_FAILED = 'RECEIVE_NEW_USER_FAILED',
+}
+
+export interface CurrentUser {
+	ID: number;
+	display_name: string;
+	username: string;
+}
+
+export interface NewUser {
+	userId: string | number | undefined;
+	username: string | undefined;
+	bearerToken: string | null | undefined;
+}
+
+export interface NewUserResponse {
+	success: boolean;
+	is_signup_sandbox: boolean;
+	username?: string;
+	signup_sandbox_username?: string;
+	user_id?: number;
+	signup_sandbox_user_id?: string;
+	bearer_token?: string | null;
+}
```

Comment:
I haven't looked the actual server response, but I'm guessing some of these properties are only optional because they don't exist when `success===false`?

In which case the data can be modeled like this:
```typescript
interface NewUserSuccessResponse {
	success: true;
	is_signup_sandbox: boolean;
	bearer_token: string | null;

	// I guess these would still have to be optional since only one is defined I think.
	username?: string;
	signup_sandbox_username?: string;
	user_id?: number;
	signup_sandbox_user_id?: string;
}

interface NewUserSuccessResponse {
	success: false;
	is_signup_sandbox: boolean;
}

export type NewUserResponse = NewUserSuccessResponse | NewUserErrorResponse;
```

That means that after checking that `success===true` you won't need to do undefined checks on all of those properties.

On the other hand, now that we have `.?` syntax maybe having all of these optional properties isn't so bothersome anymore.

Reply:
`bearer_token` should in theory always be a string too if `success === true`

Reply:
💯Beautiful. These are [_discriminated unions_](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) if anyone is interested in details.

Reply:
It turns out that an unsuccessful signup has a completely different object response and no `success` param — the `success` param _is_ included when you call the endpoint with `validate: true` where an account isn't created. So I'm not sure this discriminated union is totally necessary here, but I've left it in because it seems like it could still be useful in the future (I just haven't used it in the reducer).